### PR TITLE
Base Evaluation (TPDM) models

### PR DESCRIPTION
**Pull request by maxalmazov**

### CODE REVIEW

File: models/staging/tpdm/base/base_tpdm__evaluation_element.sql [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        pull_timestamp,
+        last_modified_timestamp,
+        file_row_number,
+        filename,
+        is_deleted,
+
+        v:evaluationElementTitle::string as evaluation_element_title,
+        v:sortOrder::int                 as sort_order,
+        v:minRating::decimal             as min_rating,
+        v:maxRating::decimal             as max_rating,
```
Comment (maxalmazov) [inline models/staging/tpdm/base/base_tpdm__evaluation_element.sql]: I'm not sure about type, what type here should be? it is DECIMAL(3,6) in docs 
image

File: models/staging/tpdm/base/base_tpdm__evaluation_objective.sql [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        last_modified_timestamp,
+        file_row_number,
+        filename,
+        is_deleted,
+
+        v:evaluationObjectiveTitle::string       as evaluation_objective_title,
+        v:evaluationObjectiveDescription::string as evaluation_objective_description,
+        v:sortOrder::int                         as sort_order,
+        v:minRating::decimal                     as min_rating,
+        v:maxRating::decimal                     as max_rating,
```
Comment (maxalmazov) [inline models/staging/tpdm/base/base_tpdm__evaluation_objective.sql]: And here the same

File: models/staging/tpdm/base/base_tpdm__evaluation.sql [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+        v:evaluationTitle::string             as evaluation_title,
+        v:evaluationDescription::string       as evaluation_description,
+        v:minRating::integer                  as min_rating,
+        v:maxRating::integer                  as max_rating,
+        v:interRaterReliabilityScore::integer as inter_rater_reliability_score,
+        -- descriptors
+        {{ extract_descriptor('v:evaluationTypeDescriptor::string') }} as evaluation_type,
+        -- unflattened lists
+        v:ratingLevels                       as v_rating_levels,
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__evaluation.sql]: Check indentation on this line

File: models/staging/tpdm/base/base_tpdm__evaluation.sql
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__evaluation.sql:1]: Make sure there `record_guid` column in each file - it's EDU pattern.

File: models/staging/tpdm/base/base_tpdm__evaluation.sql [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        -- references
+        v:performanceEvaluationReference as performance_evaluation_reference,
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__evaluation.sql]: Should we include `educationOrganizationId` column nested in the reference?

File: models/staging/tpdm/base/base_tpdm__evaluation_element.sql [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        v:evaluationElementTitle::string as evaluation_element_title,
+        v:sortOrder::int                 as sort_order,
+        v:minRating::decimal(6,3)        as min_rating,
+        v:maxRating::decimal(6,3)        as max_rating,
+        -- descriptors
+        {{ extract_descriptor('v:evaluationTypeDescriptor::string') }} as evaluation_type,
+        -- unflattened lists
+        v:ratingLevels as rating_levels,
+        -- references
+        v:evaluationObjectiveReference as evaluation_objective_reference
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__evaluation_element.sql]: Should we include `educationOrganizationId` column nested in the reference?

File: models/staging/tpdm/base/base_tpdm__evaluation.sql [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        api_year,
+        pull_timestamp,
+        last_modified_timestamp,
+        file_row_number,
+        filename,
+        is_deleted,
+
+        v:evaluationTitle::string             as evaluation_title,
+        v:evaluationDescription::string       as evaluation_description,
+        v:minRating::integer                  as min_rating,
```
Comment (holdorph) [inline models/staging/tpdm/base/base_tpdm__evaluation.sql]: should this be changed to decimal as well?

File: models/staging/tpdm/base/base_tpdm__evaluation_objective_rating.sql [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+        last_modified_timestamp,
+        file_row_number,
+        filename,
+        is_deleted,
+
+        v:comments::string as comments,
+        -- descriptors
+        {{ extract_descriptor('v:objectiveRatingLevelDescriptor::string') }} as objective_rating_level,
+        -- unflattened lists
+        v:objectiveRatingResults  as objective_rating_results,
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__evaluation_objective_rating.sql]: I see `results` list in JSON example. Should we include this?

File: models/staging/tpdm/base/base_tpdm__evaluation_rating.sql [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        -- descriptors
+        {{ extract_descriptor('v:evaluationRatingStatusDescriptor::string') }} as evaluation_rating_status,
+        {{ extract_descriptor('v:evaluationRatingLevelDescriptor::string') }}  as evaluation_rating_level,
+        -- unflattened lists
+        v:reviewers as reviewers,
+        v:results   as results,
+        -- references
+        v:evaluationReference                  as evaluation_reference,
+        v:performanceEvaluationRatingReference as performance_evaluation_rating_reference,
+        v:sectionReference                     as section_reference
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__evaluation_rating.sql]: Should we include `sectionIdentifier` columns nested in the reference to be associated with section entity?

File: models/staging/tpdm/base/base_tpdm__performance_evaluation.sql [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        {{ extract_descriptor('v:performanceEvaluationTypeDescriptor::string') }} as performance_evaluation_type,
+        {{ extract_descriptor('v:academicSubjectDescriptor::string') }}           as academic_subjec,
+        {{ extract_descriptor('v:gradeLevels::string') }}                         as grade_levels,
+        {{ extract_descriptor('v:programGateways::string') }}                     as program_gateways,
+        {{ extract_descriptor('v:evaluationPeriodDescriptor::string') }}          as evaluation_period_descriptor,
+        -- unflattened lists
+        v:ratingLevels as rating_levels,
+        -- references
+        v:educationOrganizationReference as education_organization_reference
+        v:schoolYearTypeReference        as school_year_type_reference
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__performance_evaluation.sql]: Should we include `v:schoolYearTypeReference:schoolYear::integer as school_year` from the reference?

File: models/staging/tpdm/base/base_tpdm__performance_evaluation_rating.sql [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    v:actualDuration::string as actual_duration,
+    v:announced::boolean     as actual_duration,
+    -- descriptors
+    {{ extract_descriptor('v:performanceEvaluationRatingLevelDescriptor::string') }} as performance_evaluation_rating_level,
+    {{ extract_descriptor('v:coteachingStyleObservedDescriptor::string') }}          as coteaching_style_observed,
+    -- unflattened lists
+    v:reviewers as reviewers,
+    v:results   as results,
+    -- references
+    v:personReference                as person_reference,
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__performance_evaluation_rating.sql]: Should we include `personId` column nested in the reference to be associated with person entity?

File: models/staging/tpdm/base/base_tpdm__performance_evaluation_rating.sql [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    last_modified_timestamp,
+    file_row_number,
+    filename,
+    is_deleted,
+
+    v:actualTime::timestamp  as actual_time,
+    v:scheduleDate::date     as schedule_date,
+    v:actualDate::date       as actual_date,
+    v:comments::string       as comments,
+    v:actualDuration::string as actual_duration,
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__performance_evaluation_rating.sql]: Should `actual_duration` be int data type?

File: models/staging/tpdm/base/base_tpdm__evaluation_element_rating.sql [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+        v:areaOfReinforcement::string as area_of_reinforcement,
+        v:comments::string            as comments,
+        v:feedback::string            as feedback,
+        -- descriptors
+        {{ extract_descriptor('v:evaluationElementRatingLevelDescriptor::string') }} as evaluation_element_rating_level,
+        -- unflattened lists
+        v:elementRatingResults  as element_rating_results,
+        -- references
+        v:evaluationElementReference         as evaluation_element_reference,
+        v:evaluationObjectiveRatingReference as evaluation_objective_rating_reference
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__evaluation_element_rating.sql]: Same question here, should we pull out identificators from references for association in the model?

**Review by ChristianMurphy — COMMENTED** [review]
I'd second @zaharenka's questions

File: models/staging/tpdm/base/base_tpdm__evaluation_objective_rating.sql (hunk 7 above)
Comment (maxalmazov) [reply models/staging/tpdm/base/base_tpdm__evaluation_objective_rating.sql]: I will add, but it is a question, there is not such value [in docs](https://schema.ed-fi.org/datahandbook-v510/index.html#/EvaluationObjectiveRatingc109dabd-8fdb-4a2f-85d3-1ab475ae283c)

File: models/staging/tpdm/base/base_tpdm__evaluation.sql [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+        filename,
+        is_deleted,
+
+        v:id::string                                                   as record_guid,
+        v:evaluationTitle::string                                      as evaluation_title,
+        v:evaluationDescription::string                                as evaluation_description,
+        v:minRating::decimal(6,3)                                      as min_rating,
+        v:maxRating::decimal(6,3)                                      as max_rating,
+        v:interRaterReliabilityScore::integer                          as inter_rater_reliability_score,
+        v:performanceEvaluationReference:educationOrganizationId:: int as ed_org_id
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__evaluation.sql]: Please move id's on the first places in models

Comment (zaharenka) [reply models/staging/tpdm/base/base_tpdm__evaluation.sql]:
```suggestion
        v:performanceEvaluationReference:educationOrganizationId::int as ed_org_id
```

File: models/staging/tpdm/base/base_tpdm__evaluation_element.sql [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+        file_row_number,
+        filename,
+        is_deleted,
+
+        v:id::string                                                 as record_guid,
+        v:evaluationElementTitle::string                             as evaluation_element_title,
+        v:sortOrder::int                                             as sort_order,
+        v:minRating::decimal(6,3)                                    as min_rating,
+        v:maxRating::decimal(6,3)                                    as max_rating,
+        v:evaluationObjectiveReference:educationOrganizationId ::int as ed_org_id,
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__evaluation_element.sql]:
```suggestion
        v:evaluationObjectiveReference:educationOrganizationId::int as ed_org_id,
```

Comment (zaharenka) [reply models/staging/tpdm/base/base_tpdm__evaluation_element.sql]: Place id's on the first places in CTE's, please. Check this pattern across all models you created

File: models/staging/tpdm/base/base_tpdm__evaluation_objective.sql [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+        v:sortOrder::int                                   as sort_order,
+        v:minRating::decimal(6,3)                          as min_rating,
+        v:maxRating::decimal(6,3)                          as max_rating,
+        v:evaluationReference:educationOrganizationId::int as max_rating,
+        -- descriptors
+        {{ extract_descriptor('v:evaluationTypeDescriptor::string') }} as evaluation_type,
+        -- unflattened lists
+        v:ratingLevels as rating_levels,
+        -- references
+        v:evaluationReference as evaluation_reference
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__evaluation_objective.sql]: Please, check if you pulled out id from references across all models you created

File: models/staging/tpdm/base/base_tpdm__evaluation_objective_rating.sql (hunk 7 above)
Comment (zaharenka) [reply models/staging/tpdm/base/base_tpdm__evaluation_objective_rating.sql]: Let's rely on API docs primarily. If this becomes a bug for EA, I hope the responsible person can fix it easily

File: models/staging/tpdm/base/base_tpdm__evaluation_element_rating.sql (hunk 12 above)
Comment (maxalmazov) [reply models/staging/tpdm/base/base_tpdm__evaluation_element_rating.sql]: we have org_id in both reference, which one should we pull? should these org_ids be the same? all `rating` models the same

Comment (zaharenka) [reply models/staging/tpdm/base/base_tpdm__evaluation_element_rating.sql]: I hope they are the same, but I'd like to suggest pull from `evaluationObjectiveRatingReference ` as the UML says EvaluationElementRating has associated with EvaluationObjectiveRating.
image

File: models/staging/tpdm/base/base_tpdm__evaluation_rating.sql [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+        tenant_code,
+        api_year,
+        pull_timestamp,
+        last_modified_timestamp,
+        file_row_number,
+        filename,
+        is_deleted,
+
+        v:id::string                                  as record_guid,
+        v:sectionReference:sectionIdentifier:: string as section_id,
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__evaluation_rating.sql]:
```suggestion
        v:sectionReference:sectionIdentifier::string as section_id,
```

File: models/staging/tpdm/base/base_tpdm__performance_evaluation.sql [hunk 17]
```diff
@@ -0,0 +1,33 @@
+with educator_prep_programs as (
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__performance_evaluation.sql]: CTE name isn't correct?

Comment (zaharenka) [reply models/staging/tpdm/base/base_tpdm__performance_evaluation.sql]: Looks like copy paste typo. Please check in every model this issue

File: models/staging/tpdm/base/base_tpdm__performance_evaluation_rating.sql [hunk 18]
```diff
@@ -0,0 +1,34 @@
+with educator_prep_programs as (
+    {{ source_edfi3('performance_evaluation_rating') }}
+    ),
+    renamed as (
+select
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__performance_evaluation_rating.sql]: The formatting is messed up

File: models/staging/tpdm/base/base_tpdm__performance_evaluation.sql (hunk 17 above)
Comment (holdorph) [reply models/staging/tpdm/base/base_tpdm__performance_evaluation.sql]: agreed, this needs to change in all the files.

**Review by holdorph — CHANGES_REQUESTED** [review]

File: models/staging/tpdm/base/base_tpdm__evaluation_element_rating.sql (hunk 12 above)
Comment (maxalmazov) [reply models/staging/tpdm/base/base_tpdm__evaluation_element_rating.sql]: We discussed on daily stand up to leave it as it is, we will be able to add `ed_org_id` on the next layer

File: models/staging/tpdm/base/base_tpdm__performance_evaluation.sql (hunk 17 above)
Comment (maxalmazov) [reply models/staging/tpdm/base/base_tpdm__performance_evaluation.sql]: thanks, guys, fixed

File: models/staging/tpdm/base/base_tpdm__evaluation_element_rating.sql (hunk 12 above)
Comment (zaharenka) [reply models/staging/tpdm/base/base_tpdm__evaluation_element_rating.sql]: Hmmm...but other models created by not us have it on the base layer, e.g. \models\staging\edfi_3\base\base_ef3__course_offerings.sql

Comment (zaharenka) [reply models/staging/tpdm/base/base_tpdm__evaluation_element_rating.sql]: Models created in this PR already have it on the base. Why we don't have to do the same for this one?

File: models/staging/tpdm/base/base_tpdm__evaluation.sql [hunk 19]
```diff
@@ -0,0 +1,29 @@
+with evaluation as (
+    {{ source_edfi3('evaluation') }}
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__evaluation.sql]: All source objects have plural names

Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__evaluation.sql]:
```suggestion
    {{ source_edfi3('evaluations') }}
```

File: models/staging/tpdm/base/base_tpdm__evaluation_element.sql [hunk 20]
```diff
@@ -0,0 +1,28 @@
+with evaluation_element as (
+    {{ source_edfi3('evaluation_element') }}
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__evaluation_element.sql]:
```suggestion
    {{ source_edfi3('evaluation_elements') }}
```

File: models/staging/tpdm/base/base_tpdm__evaluation_element_rating.sql [hunk 21]
```diff
@@ -0,0 +1,28 @@
+with evaluation_element_rating as (
+    {{ source_edfi3('evaluation_element_rating') }}
```
Comment (zaharenka) [inline models/staging/tpdm/base/base_tpdm__evaluation_element_rating.sql]:
```suggestion
    {{ source_edfi3('evaluation_element_ratings') }}
```

Comment (zaharenka) [reply models/staging/tpdm/base/base_tpdm__evaluation_element_rating.sql]: Please, make sure all models reference to source objects in plural

Comment (zaharenka) [reply models/staging/tpdm/base/base_tpdm__evaluation_element_rating.sql]: Also, file *.sql names

File: models/staging/tpdm/base/base_tpdm__evaluation_element_rating.sql (hunk 12 above)
Comment (maxalmazov) [reply models/staging/tpdm/base/base_tpdm__evaluation_element_rating.sql]: `EvaluationElementRating` has two associations we did not know from which one should we take org_id
lets discuss on stand up again

**Review by holdorph — APPROVED** [review]

**Review by ChristianMurphy — APPROVED** [review]

**Pull request merged**